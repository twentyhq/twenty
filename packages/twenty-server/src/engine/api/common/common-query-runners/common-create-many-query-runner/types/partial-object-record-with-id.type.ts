import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
