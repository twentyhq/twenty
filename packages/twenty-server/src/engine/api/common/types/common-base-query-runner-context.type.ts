import { type ObjectLiteral } from 'typeorm';

import { type CommonQueryRunnerOptions } from 'src/engine/api/common/interfaces/common-query-runner-options.interface';
import { type ResolverArgs } from 'src/engine/api/graphql/workspace-resolver-builder/interfaces/workspace-resolvers-builder.interface';

import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';

export type CommonBaseQueryRunnerContext<Input extends ResolverArgs> = {
  args: Input;
  options: CommonQueryRunnerOptions;
  workspaceDataSource: WorkspaceDataSource;
  repository: WorkspaceRepository<ObjectLiteral>;
  selectedFieldsResult: CommonSelectedFieldsResult;
  isExecutedByApiKey: boolean;
  roleId?: string;
  shouldBypassPermissionChecks: boolean;
};
