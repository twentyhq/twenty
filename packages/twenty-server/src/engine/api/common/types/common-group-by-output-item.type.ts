type AggregateValues = {
  [key: string]: number;
};

type GroupByDimensionValues = {
  groupByDimensionValues: string[];
};

export type CommonGroupByOutputItem = AggregateValues & GroupByDimensionValues;
