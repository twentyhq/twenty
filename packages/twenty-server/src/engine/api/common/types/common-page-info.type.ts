import { type PageInfo } from 'src/engine/api/rest/core/interfaces/rest-api-base.handler';

export type CommonPageInfo = {
  hasNextPage: NonNullable<PageInfo['hasNextPage']>;
  hasPreviousPage: NonNullable<PageInfo['hasPreviousPage']>;
  startCursor: string | null;
  endCursor: string | null;
};
