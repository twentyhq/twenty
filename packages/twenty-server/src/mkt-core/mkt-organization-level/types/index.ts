export * from './organization-level.types';
export * from './permissions.type';
