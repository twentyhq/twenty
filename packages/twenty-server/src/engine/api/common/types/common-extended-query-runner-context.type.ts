import { type ObjectLiteral } from 'typeorm';

import { type WorkspaceAuthContext } from 'src/engine/api/common/interfaces/workspace-auth-context.interface';

import { type CommonBaseQueryRunnerContext } from 'src/engine/api/common/types/common-base-query-runner-context.type';
import { type GraphqlQueryParser } from 'src/engine/api/graphql/graphql-query-runner/graphql-query-parsers/graphql-query.parser';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
import { type RolePermissionConfig } from 'src/engine/twenty-orm/types/role-permission-config';

export type CommonExtendedQueryRunnerContext = Omit<
  CommonBaseQueryRunnerContext,
  'authContext'
> & {
  authContext: WorkspaceAuthContext;
  rolePermissionConfig: RolePermissionConfig;
  repository: WorkspaceRepository<ObjectLiteral>;
  commonQueryParser: GraphqlQueryParser;
  workspaceDataSource: WorkspaceDataSource;
};
