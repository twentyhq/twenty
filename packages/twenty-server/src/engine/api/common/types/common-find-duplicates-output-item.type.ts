import { type ObjectRecord } from 'twenty-shared/types';

export type CommonFindDuplicatesOutputItem = {
  records: ObjectRecord[];
  totalCount: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
  endCursor: string | null;
};
