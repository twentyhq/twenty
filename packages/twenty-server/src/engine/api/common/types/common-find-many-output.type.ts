import { type ObjectRecord } from 'twenty-shared/types';

import { type CommonPageInfo } from 'src/engine/api/common/types/common-page-info.type';
import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';

export type CommonFindManyOutput = {
  records: ObjectRecord[];
  aggregatedValues: Record<string, number>;
  totalCount: number;
  pageInfo: CommonPageInfo;
  selectedFieldsResult: CommonSelectedFieldsResult;
};
