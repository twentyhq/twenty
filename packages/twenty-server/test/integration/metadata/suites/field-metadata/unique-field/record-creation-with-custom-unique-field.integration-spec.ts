import gql from 'graphql-tag';
import { makeGraphqlAPIRequest } from 'test/integration/graphql/utils/make-graphql-api-request.util';
import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import { updateOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/update-one-field-metadata.util';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { FieldMetadataType } from 'twenty-shared/types';

const createRecordQuery = gql`
  mutation CreateFirstRecord($data: TestRecordObjectCreateInput!) {
    createTestRecordObject(data: $data) {
      id
      uniqueTestField
    }
  }
`;

describe('create records with custom unique fields', () => {
  let createdObjectMetadataId = '';
  let uniqueFieldId = '';

  beforeEach(async () => {
    const {
      data: {
        createOneObject: { id: objectMetadataId },
      },
    } = await createOneObjectMetadata({
      input: {
        nameSingular: 'testRecordObject',
        namePlural: 'testRecordObjects',
        labelSingular: 'Test Record Object',
        labelPlural: 'Test Record Objects',
        icon: 'IconTestRecord',
      },
    });

    createdObjectMetadataId = objectMetadataId;

    const { data: createdField } = await createOneFieldMetadata({
      input: {
        name: 'uniqueTestField',
        label: 'Unique Test Field',
        type: FieldMetadataType.TEXT,
        objectMetadataId: createdObjectMetadataId,
        isUnique: true,
      },
      gqlFields: `
        id
        name
        label
        type
        isUnique
      `,
    });

    uniqueFieldId = createdField.createOneField.id;
  });

  afterEach(async () => {
    await deleteOneObjectMetadata({
      input: { idToDelete: createdObjectMetadataId },
    });
  });

  it('should not create the second record with same value on unique field', async () => {
    const firstRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'uniqueValue123',
        },
      },
    });

    expect(firstRecordResponse.body.errors).toBeUndefined();
    expect(firstRecordResponse.body.data.createTestRecordObject).toBeDefined();
    expect(
      firstRecordResponse.body.data.createTestRecordObject.uniqueTestField,
    ).toBe('uniqueValue123');

    const secondRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'uniqueValue123',
        },
      },
    });

    expect(secondRecordResponse.body.errors).toBeDefined();
    expect(secondRecordResponse.body.errors[0].message).toContain(
      'Duplicate Unique Test Field with value uniqueValue123. Please set a unique one.',
    );
    expect(secondRecordResponse.body.data.createTestRecordObject).toBeNull();
  });

  it('should create the second record with same value on updated field (previously unique)', async () => {
    const firstRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'duplicateValue',
        },
      },
    });

    expect(firstRecordResponse.body.errors).toBeUndefined();
    expect(
      firstRecordResponse.body.data.createTestRecordObject.uniqueTestField,
    ).toBe('duplicateValue');

    const { data, errors } = await updateOneFieldMetadata({
      input: {
        idToUpdate: uniqueFieldId,
        updatePayload: { isUnique: false },
      },
      gqlFields: `
        id
        name
        isUnique
      `,
    });

    expect(errors).toBeUndefined();
    expect(data.updateOneField.isUnique).toBe(false);

    const secondRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'duplicateValue',
        },
      },
    });

    expect(secondRecordResponse.body.errors).toBeUndefined();
    expect(secondRecordResponse.body.data.createTestRecordObject).toBeDefined();
    expect(
      secondRecordResponse.body.data.createTestRecordObject.uniqueTestField,
    ).toBe('duplicateValue');
  });

  it('should not create a unique index on field if records with same value already exist', async () => {
    await updateOneFieldMetadata({
      input: {
        idToUpdate: uniqueFieldId,
        updatePayload: { isUnique: false },
      },
      gqlFields: `
        id
        name
        isUnique
      `,
    });

    const firstRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'duplicatedValue',
        },
      },
    });

    expect(firstRecordResponse.body.errors).toBeUndefined();

    const secondRecordResponse = await makeGraphqlAPIRequest({
      query: createRecordQuery,
      variables: {
        data: {
          uniqueTestField: 'duplicatedValue',
        },
      },
    });

    expect(secondRecordResponse.body.errors).toBeUndefined();

    const { data, errors } = await updateOneFieldMetadata({
      input: {
        idToUpdate: uniqueFieldId,
        updatePayload: { isUnique: true },
      },
      gqlFields: `
        id
        name
        isUnique
      `,
      expectToFail: true,
    });

    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors[0].message).toContain(
      'Unique index creation failed because of unique constraint violation',
    );
  });
});
