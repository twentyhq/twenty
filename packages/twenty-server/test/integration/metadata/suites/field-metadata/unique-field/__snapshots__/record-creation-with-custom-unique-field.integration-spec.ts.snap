// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create records with custom unique fields should not create a unique index on field if records with same value already exist 1`] = `
{
  "extensions": {
    "code": "INTERNAL_SERVER_ERROR",
    "exceptionEventId": Any<String>,
    "userFriendlyMessage": "An error occurred.",
  },
  "message": "RUNNER_INTERNAL_SERVER_ERROR",
}
`;

exports[`create records with custom unique fields should not create the second record with same value on unique field 1`] = `
{
  "extensions": {
    "code": "BAD_USER_INPUT",
    "isExpected": true,
    "userFriendlyMessage": "This unique test field with value uniqueValue123 is already taken. Please choose a different value.",
  },
  "message": "Duplicate Unique Test Field with value uniqueValue123. Please set a unique one.",
  "name": "UserInputError",
}
`;
